### PR TITLE
feat: websockets support

### PR DESCRIPTION
**Pull request by FabioPinheiro**
Support DIDComm via Websockets

Support WebSocket URI as endpoints
User of Scala DID framework
Remove module 'http-utils'
Support multi endpoint in the config file
Better error handling
Configuration update
Cleanup duplicated code
Work for https://github.com/input-output-hk/atala-prism-mediator/issues/144

### CODE REVIEW

File: mediator/src/test/scala/io/iohk/atala/mediator/protocols/DiscoverFeaturesExecuterSpec.scala [hunk 1]
```diff
@@ -34,15 +36,18 @@ object DiscoverFeaturesExecuterSpec extends ZIOSpecDefault with DidAccountStubSe
       for {
         agent <- ZIO.service[MediatorAgent]
         msg <- ZIO.fromEither(plaintextDiscoverFeatureRequestMessage(bobAgent.id.did, agent.id.did))
-        result <- executer.execute(msg)
-        message <- ZIO.fromOption(result)
-        decryptedMessage <- authDecrypt(message.asInstanceOf[EncryptedMessage]).provideSomeLayer(bobAgentLayer)
-        featureDisclose <- ZIO.fromEither(decryptedMessage.asInstanceOf[PlaintextMessage].toFeatureDisclose)
+        action <- executer.program(msg)
+        // decryptedMessage <- authDecrypt(message.asInstanceOf[EncryptedMessage]).provideSomeLayer(bobAgentLayer)
```
Comment (mineme0110) [inline mediator/src/test/scala/io/iohk/atala/mediator/protocols/DiscoverFeaturesExecuterSpec.scala]: Remove the commented line

File: mediator/src/test/scala/io/iohk/atala/mediator/protocols/DiscoverFeaturesExecuterSpec.scala [hunk 2]
```diff
@@ -52,21 +57,21 @@ object DiscoverFeaturesExecuterSpec extends ZIOSpecDefault with DidAccountStubSe
       for {
         agent <- ZIO.service[MediatorAgent]
         msg <- ZIO.fromEither(plaintextDiscoverFeatureRequestMessageNoMatch(bobAgent.id.did, agent.id.did))
-        result <- executer.execute(msg)
-        message <- ZIO.fromOption(result)
-        decryptedMessage <- authDecrypt(message.asInstanceOf[EncryptedMessage]).provideSomeLayer(bobAgentLayer)
-        featureDisclose <- ZIO.fromEither(decryptedMessage.asInstanceOf[PlaintextMessage].toFeatureDisclose)
-
+        action <- executer.program(msg)
+        // decryptedMessage <- authDecrypt(message.asInstanceOf[EncryptedMessage]).provideSomeLayer(bobAgentLayer)
```
Comment (mineme0110) [inline mediator/src/test/scala/io/iohk/atala/mediator/protocols/DiscoverFeaturesExecuterSpec.scala]: Same here

File: mediator/src/test/scala/io/iohk/atala/mediator/protocols/ForwardMessageExecutorSpec.scala [hunk 3]
```diff
@@ -28,23 +30,21 @@ object ForwardMessageExecutorSpec extends ZIOSpecDefault with DidAccountStubSetu
...
-        val jsonString = Base64.fromBase64url(signedMessage.payload.base64url).decodeToString
-        val problemReport = jsonString.fromJson[PlaintextMessage].flatMap(ProblemReport.fromPlaintextMessage)
-        assert(problemReport)(
-          isRight(
-            hasField("code", (p: ProblemReport) => p.code, equalTo(ProblemCode.ErroFail("req", "not_enroll"))) &&
-              hasField(
-                "from",
-                (p: ProblemReport) => p.from,
-                equalTo(alice)
+        // TODO assertTrue(message.isInstanceOf[SignedMessage])
```
Comment (mineme0110) [inline mediator/src/test/scala/io/iohk/atala/mediator/protocols/ForwardMessageExecutorSpec.scala]: Remove the commented code

File: mediator/src/main/scala/io/iohk/atala/mediator/actions/ActionUtils.scala [hunk 4]
```diff
@@ -1,158 +1,157 @@
...
-    Operations & Agent & Resolver & MessageDispatcher & OutboxMessageRepo,
-    MediatorError,
-    Option[SignedMessage | EncryptedMessage]
-  ] =
-    action match {
-      case _: NoReply.type => ZIO.succeed(None)
-      case action: AnyReply =>
-        val reply = action.msg
-        for {
+// object ActionUtils {
```
Comment (mineme0110) [inline mediator/src/main/scala/io/iohk/atala/mediator/actions/ActionUtils.scala]: Do we still need this ? mediator/src/main/scala/io/iohk/atala/mediator/actions/ActionUtils.scala

Comment (FabioPinheiro) [reply mediator/src/main/scala/io/iohk/atala/mediator/actions/ActionUtils.scala]: I think I still need parts of this code.
But I need to run the test suite first. I don't know why the test suite is green...

**Review by mineme0110 — APPROVED** [review]

File: mediator/src/main/scala/io/iohk/atala/mediator/protocols/PickupExecuter.scala [hunk 5]
```diff
@@ -134,10 +128,14 @@ object PickupExecuter
                   messagesToReturn =
                     if (m.recipient_did.isEmpty) allMessagesFor
                     else {
-                      allMessagesFor.filterNot(
-                        _.msg.recipientsSubject
-                          .map(_.did)
-                          .forall(e => !m.recipient_did.map(_.toDID.did).contains(e))
+                      allMessagesFor.filterNot(item =>
+                        item.msg match {
+                          case sMsg: SignedMessage => ??? // FIXME TODO
```
Comment (FabioPinheiro) [inline mediator/src/main/scala/io/iohk/atala/mediator/protocols/PickupExecuter.scala]: TODO

File: README.md [hunk 6]
```diff
@@ -136,7 +136,7 @@ To set up the mediator identity:
 - `KEY_AGREEMENT_X` - is the key agreement public key (MUST be a X25519 OKP key type).
 - `KEY_AUTHENTICATION_D` - is the key authentication private key (MUST be an Ed25519 OKP key type).
 - `KEY_AUTHENTICATION_X` - is the key authentication public key (MUST be an Ed25519 OKP key type).
-- `SERVICE_ENDPOINT` - is the endpoint of the mediator. Where the mediator will be listening to incoming DID Comm messages.
+- `SERVICE_ENDPOINTS` - is the list of endpoints of the mediator split by ';'. Where the mediator will be listening to incoming DID Comm messages.
```
Comment (petevielhaber) [inline README.md]:
```suggestion
- `SERVICE_ENDPOINTS` - is the list of endpoints of the mediator split by ';' where the mediator will listen to incoming DIDComm messages.
```

**Review by petevielhaber — APPROVED** [review]
- Made minor change to grammar.

**Pull request merged**